Course description popover

singingwolfboy opened the pull request:
Fixes #1237. Follow-up to #1378.

This pull request builds on the work introduced in #1378 to implement the program course listing as a React component. In addition, we use the [Material UI Popover component](http://www.material-ui.com/#/components/popover) to display the course description when the user clicks on the course title. (Course descriptions in screenshots provided by [hipsum.co](https://hipsum.co).)

![no description visibile](https://cloud.githubusercontent.com/assets/132355/19389587/84484bf8-91f2-11e6-9d47-476857a59a10.png)
![description popover, desktop](https://cloud.githubusercontent.com/assets/132355/19389602/86c125f8-91f2-11e6-8b4b-d8226ba2ae8a.png)
![description popover, tablet](https://cloud.githubusercontent.com/assets/132355/19389606/898caed8-91f2-11e6-8fe0-c8eea62544b8.png)
![description popover, mobile](https://cloud.githubusercontent.com/assets/132355/19389608/8b45f7e8-91f2-11e6-923a-204ca846dc46.png)

aliceriot commented:
functionality looks good!

aliceriot commented:
I think these links should be `cursor: pointer` though:

![pointer_thing](https://cloud.githubusercontent.com/assets/6207644/19397975/45c8fd5c-9218-11e6-854e-715563c65ee0.png)

aliceriot commented:
taking a look at the code now

roberthouse54 commented:
Looks good, but can we lose the animated transitions?

aliceriot commented on static/js/components/ProgramCourse.js:
```diff
@@ -0,0 +1,74 @@
+      </li>
+    );
+  }
+}
```
general code formatting thing

we tend to leave a blank link between every method on a class (and every method in a file), and between the imports and any code.

so, in miniature, a typical class declaration looks like this:

```js
import { Foo } from 'bar';

class Foobar extends Foo {
  props: {
    foo: string,
  }
  
  baz () {
    // weee
  }
  
  foobar () {
    // moar code
  }
}
```

aliceriot commented on static/js/components/ProgramCourse.js:
```diff
@@ -0,0 +1,74 @@
+  }
+  state = {
+    isOpen: false,
+  }
```
we generally don't use `this.state` much (generally we use redux instead). I think it's fine here, since it's one tiny bit of state and it's outside of the main application.

aliceriot commented on static/js/components/ProgramCourse.js:
```diff
@@ -0,0 +1,74 @@
+import Popover from 'material-ui/Popover';
+
+
+export default class ProgramCourse extends React.Component {
```
I think `ProgramCourse` isn't the best name. `CourseDescription` is unfortunately already taken by a component in the dashboard.

maybe `CourseInfoPopover`? Idk, just something that better captures the functionality that this class implements.

aliceriot commented on static/js/components/ProgramCourse.js:
```diff
@@ -0,0 +1,74 @@
+
+    let titleEl, popoverLink;
+    if (url) {
+      titleEl = <a href={url}>{title}</a>;
```
will it be possible to click on `titleEl`? it looks below like it gets wrapped in an `h4` tag which has an `onclick` handler

aliceriot commented on static/js/components/ProgramCourse.js:
```diff
@@ -0,0 +1,74 @@
+        >
+          <h4 className="title">{title}</h4>
+          <div className="description course-description">{description}</div>
+          {popoverLink}
```
this is just my preference, but for the `popoverLink` it looks like it only depends on the `url`, so why not implement a little method:

```js
popoverLink = url => (
  url ? <a className="edx-link" href={url}>View on edX</a> : null
);
```

and then here in your JSX you'd just do `popoverLink(url)`.

aliceriot commented on static/js/components/ProgramCourseList.js:
```diff
@@ -0,0 +1,20 @@
+
+export default class ProgramCourseList extends React.Component {
+  props: {
+    courses: Array<Object>,
```
It would be good to make `Object` more specific, maybe with a `ProgramPageCourse` type or something like that

aliceriot commented on static/js/components/ProgramCourseList.js:
```diff
@@ -0,0 +1,20 @@
+import ProgramCourse from './ProgramCourse';
+
+
+export default class ProgramCourseList extends React.Component {
```
it might be good to try to namespace this to the program page somehow, just because we have a `CourseListCard` already and it might get a bit confusing.

OR maybe we could move these components to a new directory, like `static/js/components/program_page` or something.

aliceriot commented on static/js/components/ProgramCourseList.js:
```diff
@@ -0,0 +1,20 @@
+  render() {
+    const courseDetails = this.props.courses.map((course, index) =>
+      <ProgramCourse key={index} {...course} />
+    );
```
like above, my pref here would be a method that takes all the courses and returns this, which you then call below

aliceriot commented on static/js/program_course_list.js:
```diff
@@ -0,0 +1,20 @@
+const ThemedProgramCourseList = () => (
+  <MuiThemeProvider muiTheme={getMuiTheme()}>
+    <ProgramCourseList courses={SETTINGS.courses} />
+  </MuiThemeProvider>
```
the JSX here could be inlined in the `ReactDOM.render` call below, although it doesn't make a big difference

aliceriot commented:
is this ready for review again?

singingwolfboy commented:
@aliceriot: yes it is

aliceriot commented:
word :)

aliceriot commented on static/js/components/CourseListItemWithPopover.js:
```diff
@@ -0,0 +1,97 @@
+    );
+    const popoverLink = (url) => (
+      url ? <a className="edx-link" href={url}>View on edX</a> : null
+    );
```
I'd declare these two functions outside of `render`, they don't have any serious runtime dependency which they couldn't take as an argument.

aliceriot commented on static/js/components/CourseListWithPopover.js:
```diff
@@ -0,0 +1,24 @@
+// @flow
+import React from 'react';
+import CourseListItemWithPopover from './CourseListItemWithPopover';
+
```
we usually just leave a single blank line above the class definition (here and in the other files too)

aliceriot commented on static/js/components/CourseListWithPopover.js:
```diff
@@ -0,0 +1,24 @@
+      return courses.map((course, index) =>
+        <CourseListItemWithPopover key={index} {...course} />
+      );
+    };
```
this is another function it would be good to pull out of `render`

singingwolfboy commented:
@aliceriot: I think I've addressed your comments. The only error I'm seeing when I run `docker-compose run watch npm run-script flow` is `/src/node_modules/flow-bin/vendor/flow: 1: /src/node_modules/flow-bin/vendor/flow: Syntax error: word unexpected (expecting ")")`, which I can reproduce on the `master` branch, as well.

aliceriot commented on static/js/course_list.js:
```diff
@@ -0,0 +1,16 @@
+// @flow
+/* global SETTINGS:false */
```
can you add `__webpack_public_path__` at the top of this file? you can take a look at a bunch of the other entry points (e.g. `static/js/dashboard.js`) for an example.

singingwolfboy replied:
Done.

aliceriot commented on static/js/components/CourseListItemWithPopover.js:
```diff
@@ -0,0 +1,96 @@
+const titleEl = (title, url) => (
+  url ? <a href={url}>{title}</a> : title
+);
+const popoverLink = (url) => (
```
could you add a blank line between these?

aliceriot commented:
I'm getting some flow errors still, are you seeing these as well?

singingwolfboy commented:
@aliceriot: as I said in a previous comment, when I run `docker-compose run watch npm run-script flow`, I get `/src/node_modules/flow-bin/vendor/flow: 1: /src/node_modules/flow-bin/vendor/flow: Syntax error: word unexpected (expecting ")")`, which I can reproduce on the master branch, as well. This may be an error with the `flow` script itself, but if that's the case, I don't know how to fix it.

In the interest of continuing to make progress, can we merge this pull request as-is? I can't fix problems that I can't identify. If you are able to see flow problems that I can't, then maybe you could fix them in a follow-up pull request, and tag me so I can learn what the issues are.

aliceriot commented:
Ok, my preference would be to not merging code that isn't passing flow, but I think it's alright in this instance.
